File: WinForms/VB/GridMainDemo/Modules/ColumnCustomization.vb
Imports System
Imports System.Drawing
Imports System.ComponentModel
Imports System.Windows.Forms

Namespace DevExpress.XtraGrid.Demos

    ''' <summary>
    ''' Summary description for ColumnCustomization.
    ''' </summary>
    Public Partial Class ColumnCustomization
        Inherits TutorialControl

        Public Sub New()
            '
            ' Required for Windows Form Designer support
            '
            InitializeComponent()
            AddHandler tmr.Tick, New EventHandler(AddressOf tmr_Tick)
        '
        ' TODO: Add any constructor code after InitializeComponent call
        '
        End Sub

        Private _show As Boolean = False

        Public Overrides ReadOnly Property ExportView As Views.Base.BaseView
            Get
                Return advBandedGridView1
            End Get
        End Property

        Private Sub ColumnCustomization_Load(ByVal sender As Object, ByVal e As EventArgs)
            InitVehiclesData()
            InitCustomization()
        End Sub

        Protected Overrides ReadOnly Property WhatsThisCodeFileNames As String()
            Get
                Return New String() {"ColumnCustomization"}
            End Get
        End Property

        Protected Overrides ReadOnly Property WhatsThisXMLFileName As String
            Get
                Return "ColumnCustomization"
            End Get
        End Property

        Private tmr As Timer = New Timer()

        Protected Overrides Sub DoShow()
            MyBase.DoShow()
            _show = True
            tmr.Start()
        End Sub

        Private Sub tmr_Tick(ByVal sender As Object, ByVal e As EventArgs)
            tmr.Stop()
            If Visible Then ShowColumnSelector()
        End Sub

        Protected Overrides Sub InitXMLData(ByVal dataFileName As String)
            VehiclesData.InitXMLData(dataFileName, gridControl1)
            VehiclesData.InitColumnViewEditors(advBandedGridView1)
        End Sub

        Private initData As Boolean = False

        Private Sub InitCustomization()
            initData = True
            Try
                ccbSnapMode.Properties.SetFlags(GetType(Utils.Controls.SnapMode))
                For i As Integer = ccbSnapMode.Properties.Items.Count - 1 To 0 Step -1
                    If CInt(ccbSnapMode.Properties.Items(i).Value) < 0 OrElse CInt(ccbSnapMode.Properties.Items(i).Value) > 8 Then ccbSnapMode.Properties.Items.RemoveAt(i)
                Next

                checkEdit1.Checked = advBandedGridView1.OptionsCustomization.AllowChangeColumnParent
                checkEdit2.Checked = advBandedGridView1.OptionsCustomization.AllowChangeBandParent
                checkEdit3.Checked = advBandedGridView1.OptionsCustomization.ShowBandsInCustomizationForm
                checkEdit4.Checked = advBandedGridView1.OptionsCustomization.CustomizationFormSearchBoxVisible
                ccbSnapMode.EditValue = advBandedGridView1.OptionsCustomization.CustomizationFormSnapMode
            Finally
                initData = False
            End Try
        End Sub

        '<simpleButton1>
        Private Sub simpleButton1_Click(ByVal sender As Object, ByVal e As EventArgs)
            _show = Not _show
            ShowColumnSelector()
        End Sub

        Private Sub ShowColumnSelector()
            ShowColumnSelector(True)
        End Sub

        Private Sub ShowColumnSelector(ByVal showForm As Boolean)
            If _show Then
                simpleButton1.Text = Properties.Resources.HideColumnsSelector
                If showForm Then advBandedGridView1.ColumnsCustomization()
            Else
                simpleButton1.Text = Properties.Resources.ShowColumnsSelector
                If showForm Then advBandedGridView1.DestroyCustomization()
            End If
        End Sub

        '</simpleButton1>
        Private Sub advBandedGridView1_ShowCustomizationForm(ByVal sender As Object, ByVal e As EventArgs)
            _show = True
            ShowColumnSelector(False)
        End Sub

        Private Sub advBandedGridView1_HideCustomizationForm(ByVal sender As Object, ByVal e As EventArgs)
            _show = False
            ShowColumnSelector(False)
        End Sub

        '<checkEdit1>
        Private Sub checkEdit1_CheckedChanged(ByVal sender As Object, ByVal e As EventArgs)
            Dim edit As XtraEditors.CheckEdit = TryCast(sender, XtraEditors.CheckEdit)
            advBandedGridView1.OptionsCustomization.AllowChangeColumnParent = edit.Checked
        End Sub

        '</checkEdit1>
        '<checkEdit2>
        Private Sub checkEdit2_CheckedChanged(ByVal sender As Object, ByVal e As EventArgs)
            Dim edit As XtraEditors.CheckEdit = TryCast(sender, XtraEditors.CheckEdit)
            advBandedGridView1.OptionsCustomization.AllowChangeBandParent = edit.Checked
        End Sub

        '</checkEdit2>
        '<checkEdit3>
        Private Sub checkEdit3_CheckedChanged(ByVal sender As Object, ByVal e As EventArgs)
            Dim edit As XtraEditors.CheckEdit = TryCast(sender, XtraEditors.CheckEdit)
            advBandedGridView1.OptionsCustomization.ShowBandsInCustomizationForm = edit.Checked
        End Sub

        '</checkEdit3>
        Private Sub checkEdit4_CheckedChanged(ByVal sender As Object, ByVal e As EventArgs)
            Dim edit As XtraEditors.CheckEdit = TryCast(sender, XtraEditors.CheckEdit)
            advBandedGridView1.OptionsCustomization.CustomizationFormSearchBoxVisible = edit.Checked
        End Sub

#Region "ReportGeneration"
        Public Overrides ReadOnly Property AllowGenerateReport As Boolean
            Get
                Return False
            End Get
        End Property

#End Region
        Private Sub ccbSnapMode_EditValueChanged(ByVal sender As Object, ByVal e As EventArgs)
            If initData Then Return
            Dim edit As XtraEditors.CheckedComboBoxEdit = TryCast(sender, XtraEditors.CheckedComboBoxEdit)
            If Not String.IsNullOrEmpty(edit.Text) Then advBandedGridView1.OptionsCustomization.CustomizationFormSnapMode = CType(edit.EditValue, Utils.Controls.SnapMode)
        End Sub

        Private Sub layoutControlItem6_CustomDraw(ByVal sender As Object, ByVal e As XtraLayout.ItemCustomDrawEventArgs)
            e.DefaultDraw()
            e.Cache.FillRectangle(Color.FromArgb(50, 127, 221, 134), e.Bounds)
            e.Handled = True
        End Sub
    End Class
End Namespace

File: WinForms/CS/GridMainDemo/Modules/ColumnCustomization.cs
using System;
using System.Drawing;
using System.Collections;
using System.ComponentModel;
using System.Windows.Forms;
using System.Data;

namespace DevExpress.XtraGrid.Demos {
    /// <summary>
    /// Summary description for ColumnCustomization.
    /// </summary>
    public partial class ColumnCustomization : TutorialControl {
        public ColumnCustomization() {
            //
            // Required for Windows Form Designer support
            //
            InitializeComponent();
            tmr.Tick += new EventHandler(tmr_Tick);
            //
            // TODO: Add any constructor code after InitializeComponent call
            //
        }
        bool _show = false;
        public override DevExpress.XtraGrid.Views.Base.BaseView ExportView { get { return advBandedGridView1; } }
        private void ColumnCustomization_Load(object sender, System.EventArgs e) {
            InitVehiclesData();
            InitCustomization();
        }
        protected override string[] WhatsThisCodeFileNames {
            get { return new string[] { "ColumnCustomization" }; }
        }
        protected override string WhatsThisXMLFileName {
            get { return "ColumnCustomization"; }
        }

        Timer tmr = new Timer();
        protected override void DoShow() {
            base.DoShow();
            _show = true;
            tmr.Start();
        }

        void tmr_Tick(object sender, EventArgs e) {
            tmr.Stop();
            if(Visible)
                ShowColumnSelector();
        }
        protected override void InitXMLData(string dataFileName) {
            VehiclesData.InitXMLData(dataFileName, gridControl1);
            VehiclesData.InitColumnViewEditors(advBandedGridView1);
        }
        bool initData = false;
        private void InitCustomization() {
            initData = true;
            try {
                ccbSnapMode.Properties.SetFlags(typeof(DevExpress.Utils.Controls.SnapMode));
                for(int i = ccbSnapMode.Properties.Items.Count - 1; i >= 0; i--) {
                    if((int)ccbSnapMode.Properties.Items[i].Value < 0 || (int)ccbSnapMode.Properties.Items[i].Value > 8)
                        ccbSnapMode.Properties.Items.RemoveAt(i);
                }

                checkEdit1.Checked = advBandedGridView1.OptionsCustomization.AllowChangeColumnParent;
                checkEdit2.Checked = advBandedGridView1.OptionsCustomization.AllowChangeBandParent;
                checkEdit3.Checked = advBandedGridView1.OptionsCustomization.ShowBandsInCustomizationForm;
                checkEdit4.Checked = advBandedGridView1.OptionsCustomization.CustomizationFormSearchBoxVisible;
                ccbSnapMode.EditValue = advBandedGridView1.OptionsCustomization.CustomizationFormSnapMode;
            } finally {
                initData = false;
            }
        }
        //<simpleButton1>
        private void simpleButton1_Click(object sender, System.EventArgs e) {
            _show = !_show;
            ShowColumnSelector();
        }
        
        private void ShowColumnSelector() { ShowColumnSelector(true); }
        private void ShowColumnSelector(bool showForm) {
            if(_show) {
                simpleButton1.Text = Properties.Resources.HideColumnsSelector;
                if(showForm) advBandedGridView1.ColumnsCustomization();
            }
            else {
                simpleButton1.Text = Properties.Resources.ShowColumnsSelector;
                if(showForm) advBandedGridView1.DestroyCustomization();
            }
        }
        //</simpleButton1>
        private void advBandedGridView1_ShowCustomizationForm(object sender, System.EventArgs e) {
            _show = true;
            ShowColumnSelector(false);
        }

        private void advBandedGridView1_HideCustomizationForm(object sender, System.EventArgs e) {
            _show = false;
            ShowColumnSelector(false);
        }
        //<checkEdit1>
        private void checkEdit1_CheckedChanged(object sender, System.EventArgs e) {
            DevExpress.XtraEditors.CheckEdit edit = sender as DevExpress.XtraEditors.CheckEdit;
            advBandedGridView1.OptionsCustomization.AllowChangeColumnParent = edit.Checked;
        }
        //</checkEdit1>
        //<checkEdit2>
        private void checkEdit2_CheckedChanged(object sender, System.EventArgs e) {
            DevExpress.XtraEditors.CheckEdit edit = sender as DevExpress.XtraEditors.CheckEdit;
            advBandedGridView1.OptionsCustomization.AllowChangeBandParent = edit.Checked;
        }
        //</checkEdit2>
        //<checkEdit3>
        private void checkEdit3_CheckedChanged(object sender, System.EventArgs e) {
            DevExpress.XtraEditors.CheckEdit edit = sender as DevExpress.XtraEditors.CheckEdit;
            advBandedGridView1.OptionsCustomization.ShowBandsInCustomizationForm = edit.Checked;
        }
        //</checkEdit3>
        private void checkEdit4_CheckedChanged(object sender, EventArgs e) {
            DevExpress.XtraEditors.CheckEdit edit = sender as DevExpress.XtraEditors.CheckEdit;
            advBandedGridView1.OptionsCustomization.CustomizationFormSearchBoxVisible = edit.Checked;
        }
        #region ReportGeneration
        public override bool AllowGenerateReport { get { return false; } }
        #endregion

        private void ccbSnapMode_EditValueChanged(object sender, EventArgs e) {
            if(initData) return;
            DevExpress.XtraEditors.CheckedComboBoxEdit edit = sender as DevExpress.XtraEditors.CheckedComboBoxEdit;
            if(!string.IsNullOrEmpty(edit.Text))
                advBandedGridView1.OptionsCustomization.CustomizationFormSnapMode = (DevExpress.Utils.Controls.SnapMode)edit.EditValue;
        }

        private void layoutControlItem6_CustomDraw(object sender, XtraLayout.ItemCustomDrawEventArgs e) {
            e.DefaultDraw();
            e.Cache.FillRectangle(Color.FromArgb(50, 127, 221, 134), e.Bounds);
            e.Handled = true;
        }
    }
}
